src/main/java/io/airlift/slice/SliceUtf8.java

```diff
@@ -0,0 +1,810 @@
+        if (!isContinuationByte(forthByte)) {
+            return -3;
+        }
+
+        if (length == 4) {
+            // 1111_0xxx 10xx_xxxx 10xx_xxxx 10xx_xxxx
+            int codePoint = ((firstByte & 0b0000_0111) << 18) |
+                    ((secondByte & 0b0011_1111) << 12) |
+                    ((thirdByte & 0b0011_1111) << 6) |
+                    (forthByte & 0b0011_1111);
+            // 4 byte code points have a limited valid range
+            if (codePoint < 0x11_0000) {
+                return codePoint;
+            }
+            return -4;
+        }
+
+        byte fifthByte = utf8.getByteUnchecked(position + 4);
+        if (!isContinuationByte(fifthByte)) {
+            return -4;
+        }
+
+        if (length == 5) {
+            // According to RFC3629 limited to 4 bytes so 5 and 6 bytes are illegal
+            return -5;
+        }
+
+        byte sixthByte = utf8.getByteUnchecked(position + 5);
+        if (!isContinuationByte(sixthByte)) {
+            return -5;
+        }
+
+        if (length == 6) {
+            // According to RFC3629 limited to 4 bytes so 5 and 6 bytes are illegal
+            return -6;
+        }
+
+        // for longer sequence
+        return -1;
+    }
+
+    static int lengthOfCodePointFromStartByteSafe(byte startByte)
+    {
+        int unsignedStartByte = startByte & 0xFF;
+        if (unsignedStartByte < 0b1000_0000) {
+            // normal ASCII
+            // 0xxx_xxxx
+            return 1;
+        }
+        if (unsignedStartByte < 0b1100_0000) {
+            // illegal bytes
+            // 10xx_xxxx
+            return -1;
+        }
+        if (unsignedStartByte < 0b1110_0000) {
+            // 110x_xxxx 10xx_xxxx
+            return 2;
+        }
+        if (unsignedStartByte < 0b1111_0000) {
+            // 1110_xxxx 10xx_xxxx 10xx_xxxx
+            return 3;
+        }
+        if (unsignedStartByte < 0b1111_1000) {
+            // 1111_0xxx 10xx_xxxx 10xx_xxxx 10xx_xxxx
+            return 4;
+        }
+        if (unsignedStartByte < 0b1111_1100) {
+            // 1111_10xx 10xx_xxxx 10xx_xxxx 10xx_xxxx 10xx_xxxx
+            return 5;
+        }
+        if (unsignedStartByte < 0b1111_1110) {
+            // 1111_110x 10xx_xxxx 10xx_xxxx 10xx_xxxx 10xx_xxxx 10xx_xxxx
+            return 6;
+        }
+        return -1;
+    }
+
+    /**
+     * Finds the index of the first byte of the code point at a position, or
+     * {@code -1} if the position is not withing the slice.
+     * </p>
+     * Note: This method does not explicitly check for valid UTF-8, and may
+     * return incorrect results or throw an exception for invalid UTF-8.
+     */
+    public static int offsetOfCodePoint(Slice utf8, int codePointPosition)
+    {
+        return offsetOfCodePoint(utf8, 0, codePointPosition);
+    }
+
+    /**
+     * Finds the index of the first byte of the code point at a position, or
+     * {@code -1} if the position is not withing the slice.
```

Comment:
Maybe ask some one else to read it? It definitely looks better. But I won't be able to tell if it's clear enough because I know what this function is doing.